src/RayTracer/Public/js/test/drag_drop.test.js

```diff
@@ -0,0 +1,70 @@
+const chai = require('chai');
+const expect = chai.expect;
+
+var drag_drop = require('../drag_drop.js');
+var form = require('../form.js');
+
+
+describe('Drag and drop', function () {
+
+    it('Starts object drag', function () {
+      drag_drop.startObjectDrag("test")
+      expect(drag_drop.new_object).to.equal("test");
+      expect(drag_drop.dragged_id).to.equal("object0");
+      drag_drop.startObjectDrag("test2")
+      expect(drag_drop.new_object).to.equal("test2");
+      expect(drag_drop.dragged_id).to.equal("object1");
+    });
+
+    it('Clears dragged object', function () {
+      drag_drop.clearObjectDrag()
+      expect(drag_drop.new_object).to.equal("");
+      expect(drag_drop.dragged_id).to.equal("");
+    });
+
+    it('Finds id in JSON array', function () {
+      form.objectsJSON = [{ id: "not this one"},{ id: "not this one"},{ id: "not this one"},{ id: "test"},{ id: "not this one"}]
+      const res = drag_drop.findObjectInJSON("test")
+      expect(res).to.equal('3');
+    });
+
+    it('Handles search in undefined', function () {
+      form.objectsJSON = undefined
+      const res = drag_drop.findObjectInJSON("test")
+      expect(res).to.equal(-1);
+    });
+
+    it('Handles search in empty', function () {
+      form.objectsJSON = []
+      const res = drag_drop.findObjectInJSON("test")
+      expect(res).to.equal(-1);
+    });
+
+    it('Adds rect to JSON', function () {
+      drag_drop.addToJSON("rect","id",100,100)
+      expect(form.objectsJSON[0].id).to.equal("id");
+    });
+
+    it('Handles adding duplicate id', function () {
+      drag_drop.addToJSON("rect","id",)
+      expect(form.objectsJSON.length).to.equal(1);
+    });
+
+    it('Adds sphere to JSON', function () {
+      drag_drop.addToJSON("sphere","id2",100,100)
```

Comment:
z value has to be added to Json for 3d?

Reply:
Hey, 
yeh that's a good point. I looked into it and the 2nd value I pass in is actually treated as both the x and y coordinates. Here's the code for it:
`"point":{"x":x,"y":yz,"z":yz},`

The reason for this is that method is used to create the object in the SVG when it's dragged in initially. The 'drag in' always happens in one of the 2D views so you never know both the Z and the Y - always just one at a time. Therefore, I set the value of z and y to be wherever the cursor entered the svg and then I keep updating the z **or** the y as the user drags the object around. Since that wasn't clear from the code (nor to me...who wrote the code hah) I'll add a comment explaining that :) 

Thanks!